scripts/us_hud/income/process.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+'''Generates cleaned CSVs for HUD Income Limits data.
+
+Produces: 
+* csv/output_[YEAR].csv
+
+Usage:
+python3 process.py
+'''
+import csv
+import datetime
+import os
+import pandas as pd
+
+URL_PREFIX = 'https://www.huduser.gov/portal/datasets/il/il'
+OUTPUT_DIR = 'csv'
+
+
+def get_url(year):
+    '''Return xls url for year.
+
+  Args:
+    year: Input year.
+
+  Returns:
+    xls url for given year.
+  '''
+    if year < 2006:
+        return ''
+    suffix = str(year)[-2:]
+    if year >= 2016:
+        return f'{URL_PREFIX}{suffix}/Section8-FY{suffix}.xlsx'
+    elif year == 2015:
+        return f'{URL_PREFIX}15/Section8_Rev.xlsx'
+    elif year == 2014:
+        return f'{URL_PREFIX}14/Poverty.xls'
+    elif year == 2011:
+        return f'{URL_PREFIX}11/Section8_v3.xls'
+    elif year >= 2009:
+        return f'{URL_PREFIX}{suffix}/Section8.xls'
+    elif year == 2008:
+        return f'{URL_PREFIX}08/Section8_FY08.xls'
+    elif year == 2007:
+        return f'{URL_PREFIX}07/Section8-rev.xls'
+    elif year == 2006:
+        return f'{URL_PREFIX}06/Section8FY2006.xls'
+    else:
+        return ''
+
+
+def compute_150(df, person):
+    '''Compute 150th percentile income in-place.
+
+  Args:
+    df: Input dataframe (will be modified).
+    person: Number of people in household.
+  '''
+    df[f'l150_{person}'] = df.apply(
+        lambda x: round(x[f'l80_{person}'] / 80 * 150), axis=1)
+
+
+def process(year, matches, output_dir):
+    '''Generate cleaned CSV.
+
+  Args:
+    year: Input year.
+    matches: Map of fips dcid -> city dcid.
+    output_dir: Directory to write cleaned CSV.
+  '''
+    url = get_url(year)
+    try:
+        df = pd.read_excel(url)
+    except:
+        print(f'No file found for {url}.')
+        return
```

Comment:
Should we fail hard here if url doesn't exist?

Reply:
since we're processing each year separately, don't fully fail so we can still try generating stats for other years. hud sometimes publishes revisions which have different urls, so it's possible certain years might change slightly